src/DoubleShapeletPsfApprox.cc

```diff
@@ -0,0 +1,467 @@
+    // Normalize to unit flux and unit circle moments.
+    result.normalize();
+    result.transformInPlace(result.evaluate().computeMoments().getGridTransform());
+    return result;
+}
+
+void DoubleShapeletPsfApproxAlgorithm::fitMoments(
+    shapelet::MultiShapeletFunction & result,
+    Control const & ctrl,
+    afw::detection::Psf::Image const & psfImage
+) {
+    MomentsFunctor func;
+    applyPixelFunctor(psfImage, func);
+    auto ellipse = func.getEllipse();
+    auto sum = func.getSum();
+    try {
+        ellipse.getCore().normalize();
+    } catch (pex::exceptions::InvalidParameterError & err) {
+        throw LSST_EXCEPT(
+            meas::base::MeasurementError,
+            err.what(),
+            INVALID_MOMENTS
+        );
+    }
+    result.transformInPlace(ellipse.getGridTransform().invert());
+    result.normalize(sum);
+    afw::geom::ellipses::Axes axesInner(result.getComponents()[0].getEllipse().getCore());
+    afw::geom::ellipses::Axes axesOuter(result.getComponents()[1].getEllipse().getCore());
+    if (!(axesInner.getB() >= ctrl.minRadius)) {  // compare for opposite to catch NaNs
+        throw LSST_EXCEPT(
+            meas::base::MeasurementError,
+            (boost::format("Semi-minor radius derived from moments (%g) is less than lower bound (%g).")
+             % axesOuter.getB() % ctrl.minRadius).str(),
+            INVALID_MOMENTS
+        );
+    }
+    double const maxRadius = ctrl.maxRadiusBoxFraction*std::sqrt(psfImage.getBBox().getArea());
+    if (!(axesOuter.getA() <= maxRadius)) {
+        throw LSST_EXCEPT(
+            meas::base::MeasurementError,
+            (boost::format("Semi-major radius derived from moments (%g) is greater than upper bound (%g).")
+             % axesOuter.getA() % maxRadius).str(),
+            INVALID_MOMENTS
+        );
+    }
+    if (!(axesOuter.getDeterminantRadius() - axesInner.getDeterminantRadius() >= ctrl.minRadiusDiff)) {
+        throw LSST_EXCEPT(
+            meas::base::MeasurementError,
+            (boost::format("Different between radii (%g - %g = %g) is smaller than lower bound (%g).")
+             % axesOuter.getDeterminantRadius() % axesInner.getDeterminantRadius()
+             % (axesOuter.getDeterminantRadius() - axesInner.getDeterminantRadius())
+             % ctrl.minRadiusDiff).str(),
+            INVALID_MOMENTS
+        );
+    }
+}
+
+namespace {
+
+// An OptimizerObjective that fits the profile of a DoubleShapelet model
+// (see DoubleShapeletPsfApproxAlgorithm::makeObjective for more info.)
+class ProfileObjective : public OptimizerObjective {
+public:
+
+    ProfileObjective(
+        afw::geom::ellipses::Ellipse const & moments,
+        DoubleShapeletPsfApproxControl const & ctrl,
+        afw::image::Image<Scalar> const & image
+    ) : OptimizerObjective(image.getBBox().getArea(), 4),
+        _minRadius(ctrl.minRadius),
+        _maxRadius(ctrl.maxRadiusBoxFraction * std::sqrt(this->dataSize)),
```

Comment:
No, this is just intended to be a geometric mean of the width and height.  But I do think the box will be close enough to square that I'm not concerned about the difference between geometric and arithmetic mean (there really isn't a good reason to pick one or the other).